toolchain/sem_ir/type_info.cpp

```diff
@@ -76,46 +76,55 @@ auto InitRepr::ForType(const File& file, TypeId type_id) -> InitRepr {
   }
 }
 
-auto NumericTypeLiteralInfo::ForType(const File& file, ClassType class_type)
+auto NumericTypeLiteralInfo::ForType(const File& file,
+                                     NameId parent_scope_name_id,
+                                     ClassType class_type)
     -> NumericTypeLiteralInfo {
-  // Quickly rule out any class that's not a specific.
-  if (!class_type.specific_id.has_value()) {
-    return NumericTypeLiteralInfo::Invalid;
-  }
-
-  // The class must be declared in the `Core` package.
-  const auto& class_info = file.classes().Get(class_type.class_id);
-  if (!file.name_scopes().IsInCorePackageRoot(class_info.scope_id)) {
-    return NumericTypeLiteralInfo::Invalid;
-  }
-
   // The class's name must be the name corresponding to a type literal.
+  const auto& class_info = file.classes().Get(class_type.class_id);
   auto name_ident = file.names().GetAsStringIfIdentifier(class_info.name_id);
   if (!name_ident) {
     return NumericTypeLiteralInfo::Invalid;
   }
-  Kind kind = llvm::StringSwitch<Kind>(*name_ident)
-                  .Case("Int", Int)
-                  .Case("UInt", UInt)
-                  .Case("Float", Float)
-                  .Default(None);
-  if (kind == None) {
-    return NumericTypeLiteralInfo::Invalid;
-  }
 
-  // There must be exactly one argument.
-  const auto& specific = file.specifics().Get(class_type.specific_id);
-  auto args = file.inst_blocks().Get(specific.args_id);
-  if (args.size() != 1) {
-    return NumericTypeLiteralInfo::Invalid;
+  if (!parent_scope_name_id.has_value()) {
+    // Quickly rule out any class that's not a specific.
+    if (!class_type.specific_id.has_value()) {
+      return NumericTypeLiteralInfo::Invalid;
+    }
+    Kind kind = llvm::StringSwitch<Kind>(*name_ident)
+                    .Case("Int", Int)
+                    .Case("UInt", UInt)
+                    .Case("Float", Float)
+                    .Default(None);
+    if (kind == None) {
+      return NumericTypeLiteralInfo::Invalid;
+    }
+
+    // There must be exactly one argument.
+    const auto& specific = file.specifics().Get(class_type.specific_id);
+    auto args = file.inst_blocks().Get(specific.args_id);
+    if (args.size() != 1) {
+      return NumericTypeLiteralInfo::Invalid;
+    }
+
+    // And the argument must be an integer value.
+    auto width_arg = file.insts().TryGetAs<IntValue>(args[0]);
+    if (!width_arg) {
+      return NumericTypeLiteralInfo::Invalid;
+    }
+    return {.kind = kind, .bit_width_id = width_arg->int_id};
   }
 
-  // And the argument must be an integer value.
-  auto width_arg = file.insts().TryGetAs<IntValue>(args[0]);
-  if (!width_arg) {
-    return NumericTypeLiteralInfo::Invalid;
+  auto parent_name_ident =
+      file.names().GetAsStringIfIdentifier(parent_scope_name_id);
+  if (parent_name_ident == "CppCompat") {
+    return llvm::StringSwitch<NumericTypeLiteralInfo>(*name_ident)
+        .Case("Long32", {Int, IntId::MakeRaw(32)})
```

Comment:
This shouldn't map to a `NumericTypeLiteralInfo` with value (`Int`, 32) because that means `i32`, not some other type that's also a 32-bit integer. We also don't have a literal syntax for `Core.Long32`, so treating it as a numeric type literal is probably not the right approach.

Let's add a `TypeLiteralInfo::Kind` value for it instead. We should also rename `TypeLiteralInfo` to something more like `RecognizedTypeInfo`, since we're using it for cases that aren't literals, such as `Core.CppCompat.NullptrT` and now `Core.CppCompat.[U]Long*`, but best not to do that in this PR I think.